test/xds/xds_client_ack_nack_test.go

```diff
@@ -87,13 +99,25 @@ func (s) TestClientResourceVersionAfterStreamRestart(t *testing.T) {
 				ackVersionsMap[id] = make(map[string]string)
 			}
 			ackVersionsMap[id][req.GetTypeUrl()] = req.GetVersionInfo()
-			if seenAllACKs(ackVersionsMap) {
-				acksReceived.Fire()
+			// Prior to stream restart, we are interested only in non-empty
+			// resource versions. The xdsclient first sends out requests with an
+			// empty version string. After receipt of requested resource, it
+			// sends out another request for the same resource, but this time
+			// with a non-empty version string, to serve as an ACK.
+			if seenAllACKs(ackVersionsMap[idBeforeRestart], true) {
+				acksReceivedBeforeRestart.Fire()
+			}
+			// After stream restart, we expect the xdsclient to send out
+			// requests with version string set to the previously ACKed
+			// versions. If it sends out requests with empty version string, it
+			// is a bug and we want this test to catch it.
+			if seenAllACKs(ackVersionsMap[idAfterRestart], false) {
```

Comment:
Why is wantNonEmpty false here. The comment above states "If it sends out requests with empty version string, it is a bug and we want this test to catch it". The documentation for ackVersionsMap says "Map from stream id to a map of resource type to resource version." However, you are ignoring the resource version, which is the whole point of this test?

Reply:
I don't think I understand the second half of your comment.

Yes, before the restart, all we care about is whether we have received ACKs for all resources. And by ACKs here, I mean non-empty version strings. At this point, we cannot compare the received versions with an expected version map, because it depends on what version the server sends. The client simply replays the version number received from the server as the ACK.

And, after restart, we want to capture the first resource version for all resources. We want this to match the versions received prior to restart. If we don't capture the first resource version and instead capture the last resource versions sent out by the client, the test will always pass (because the resource has not changed on the server, which means that the server will end up sending the same version number and the client will ACK by replaying that same version number).

The whole point of having the client send previously ACKed version numbers after stream restart is to avoid the server from having to resend all resources. This becomes specially important in deployments which have a lot of resources, and where servers handle lots of clients.

Reply:
Thanks for the explanation. Yeah, I get all of what you're saying. What I'm trying to say is that if you send true to that function, it would still fail correctly in the case of the new responses having empty version string (since if it sends empty version string it would timeout), and continue to do the check in main goroutine if the responses do contain version string and make sure there. It doesn't really matter haha but was just wondering.

Reply:
Earlier behavior of the xdsclient wrt to a single resource (this is the buggy version):
T0: ADS stream starts
T1: client requests resource A (version = "")
T2: server responds with resource A (version = "v1")
T3: client acks resource A (version = "v1")
T4: stream breaks
T5: client requests resource A (version = "")
T6: server responds with resource A (version = "v1")
T7: client acks resource A (version = "v1")

New behavior of xdsclient:
T0: ADS stream starts
T1: client requests resource A (version = "")
T2: server responds with resource A (version = "v1")
T3: client acks resource A (version = "v1")
T4: stream breaks
T5: client requests resource A (version = "v1")
<<< server knows that the client already acked "v1", so, it will not send any response >>>

If I call `seenAllACKs(ackVersionsMap[idAfterRestart], true)` from here, the test will not catch the scenario that I want it to catch, because client will eventually re-receive the same resource and will ack with version "v1" which is what will get captured in our map. Hope this helps.

Reply:
Oh duh haha thanks for the explanation. I just wanna seem proactive/involved is all ;P.